nitransforms/surface.py

```diff
@@ -0,0 +1,526 @@
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Surface transforms."""
+import pathlib
+import warnings
+import h5py
+import numpy as np
+import nibabel as nb
+from scipy import sparse
+from scipy.spatial import KDTree
+from scipy.spatial.distance import cdist
+from nitransforms.base import (
+    SurfaceMesh
+)
+
+
+class SurfaceTransformBase():
+    """Generic surface transformation class"""
+
+    def __init__(self, reference, moving, spherical=False):
+        """Instantiate a generic surface transform."""
+        if spherical:
+            if not reference.check_sphere():
+                raise ValueError("reference was not spherical")
+            if not moving.check_sphere():
+                raise ValueError("moving was not spherical")
+            reference.set_radius()
+            moving.set_radius()
+        self._reference = reference
+        self._moving = moving
+
+    def __eq__(self, other):
+        ref_coords_eq = np.all(self.reference._coords == other.reference._coords)
+        ref_tris_eq = np.all(self.reference._triangles == other.reference._triangles)
+        mov_coords_eq = np.all(self.moving._coords == other.moving._coords)
+        mov_tris_eq = np.all(self.moving._triangles == other.moving._triangles)
+        return ref_coords_eq & ref_tris_eq & mov_coords_eq & mov_tris_eq
+
+    def __invert__(self):
+        return self.__class__(self._moving, self._reference)
+
+    @property
+    def reference(self):
+        return self._reference
+
+    @reference.setter
+    def reference(self, surface):
+        self._reference = SurfaceMesh(surface)
+
+    @property
+    def moving(self):
+        return self._moving
+
+    @moving.setter
+    def moving(self, surface):
+        self._moving = SurfaceMesh(surface)
+
+    @classmethod
+    def from_filename(cls, reference_path, moving_path):
+        """Create an Surface Index Transformation from a pair of surfaces with corresponding
+         vertices."""
+        reference = SurfaceMesh(nb.load(reference_path))
+        moving = SurfaceMesh(nb.load(moving_path))
+        return cls(reference, moving)
+
+
+class SurfaceCoordinateTransform(SurfaceTransformBase):
+    """Represents surface transformations in which the indices correspond and the coordinates
+     differ. This could be two surfaces representing difference structures from the same
+     hemisphere, like white matter and pial, or it could be a sphere and a deformed sphere that
+     moves those coordinates to a different location."""
+
+    __slots__ = ("_reference", "_moving")
+
+    def __init__(self, reference, moving):
+        """Instantiate a transform between two surfaces with corresponding vertices.
+         Parameters
+        ----------
+        reference: surface
+            Surface with the destination coordinates for each index.
+        moving: surface
+            Surface with the starting coordinates for each index.
```

Comment:
I feel this is defined in reverse.
```suggestion
        reference: surface
            Surface with the starting coordinates for each index.
        moving: surface
            Surface with the destination coordinates for each index.
```